### PR TITLE
Adding a RaySGD and DGL ( Deep Graph Library) integration example(gat…

### PR DESCRIPTION
**Pull request by YeahNew**
We are currently using DGL（https://github.com/dmlc/dgl） to do graph neural network research and see that raySGD has good support for distributed DL training, so we tried to combine DGL's model training with raySGD and succeeded. Under the same training, almost the same accuracy rate was finally obtained. We want add this case as an application examplet to the project.

### CODE REVIEW

**Comment by amogkam** [comment]
@YeahNew thanks for making the PR. Can you run the linter to format the code?

pip install -r ray/python/requirements_linters.txt

bash ray/scripts/format.sh

**Comment by YeahNew** [comment]
> @YeahNew thanks for making the PR. Can you run the linter to format the code?
> 
> pip install -r ray/python/requirements_linters.txt
> 
> bash ray/scripts/format.sh

Hi, @amogkam 
I have successfully formatted the code with using the format.sh.
But I used my other account (SHUIGUYU) on another computer, YeanNew and SHUIGUYU are the same person. When using git bash, I didn't notice that the SHUIGUYU account was submitting. I'm sorry that this may have caused you trouble before.
Please take a look at the code?

**Comment by YeahNew** [comment]
@amogkam @richardliaw 
Note: 
Because Ray will set OMP_NUM_THREADS=1 by default when running, this will cause data sampling and loading to the GPU in the train_epoch method to become very slow. Therefore, it is necessary to manually set this parameter to the number of cpu cores of the machine, so that the training time of each epoch will be reduced a lot.

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
+import os
+import torch
+import time
+import numpy as np
+import argparse
+from dgl.dataloading import NodeCollator
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:6]: Can you group imports just for better readability.

```
import argparse
import os
import numpy as np
import time

import torch
import torch.nn as nn
import torch.nn.functional as F

import ray
from ray.util.sgd import TorchTrainer
from ray.util.sgd.utils import AverageMeterCollection
from ray.util.sgd.torch import TrainingOperator

import dgl
from dgl.dataloading import NodeCollator
from dgl.data import RedditDataset
from dgl.nn.pytorch import GATConv
from torch.utils.data import DataLoader

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+from dgl.nn.pytorch import GATConv
+from torch.utils.data import DataLoader
+
+print("Current Path: " + os.getcwd())
+torch.manual_seed(42)
+
+
+# define the model class
+class GAT(nn.Module):
+    def __init__(self, in_feats, n_hidden, n_classes, n_layers, n_heads,
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:25]: Could you add a docstring here describing what these args are? It'll help people use the model

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        type=int,
+        default=8265,
+        help="The port to bind the dashboard server to.")
+    args = argparser.parse_args()
+    ray.init(
+        dashboard_host=args.dashboard_host, dashboard_port=args.dashboard_port)
+    ###connect to started ray cluster
+    # ray.init(address='auto', _redis_password='5241590000000000')
+    start_time = time.time()
+    run(num_workers=3, use_gpu=True, num_epochs=args.num_epochs)
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:274]: Don't you have to pass in `args.num_workers` here? Also could you take in `use_gpu` as an arg as well?

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+    argparser.add_argument(
+        '--dashboard-port',
+        type=int,
+        default=8265,
+        help="The port to bind the dashboard server to.")
+    args = argparser.parse_args()
+    ray.init(
+        dashboard_host=args.dashboard_host, dashboard_port=args.dashboard_port)
+    ###connect to started ray cluster
+    # ray.init(address='auto', _redis_password='5241590000000000')
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:272]: You can take in the address as an arg that defaults to None. And then you can just have a single ray.init like `ray.init(address=args.address, ...)`

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        '--dashboard-host',
+        type=str,
+        default='127.0.0.1',
+        help="The host to bind the dashboard server to.")
+    argparser.add_argument(
+        '--dashboard-port',
+        type=int,
+        default=8265,
+        help="The port to bind the dashboard server to.")
+    args = argparser.parse_args()
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:268]: Looks like these `args` are accessed directly inside the training code. I would recommend passing each arg explicitly as function parameters to your `run` method instead.

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    trainer = TorchTrainer(
+        training_operator_cls=CustomTrainingOperator,
+        num_workers=num_workers,
+        config={
+            "lr": 1e-3,
+            "batch_size": 1024,
+        },
+        use_gpu=use_gpu,
+        backend="nccl",
+        # use_tqdm=True,
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:229]: remove?

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        g.ndata['labels'] = g.ndata['label']
+        self.in_feats = g.ndata['features'].shape[1]
+        self.n_classes = data.num_classes
+        # add self loop,
+        g = dgl.remove_self_loop(g)
+        g = dgl.add_self_loop(g)
+        # Create csr/coo/csc formats before launching training processes
+        g.create_formats_()
+        self.g = g
+        train_nid = th.nonzero(g.ndata['train_mask'], as_tuple=True)[0]
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:95]: Would you be able to use `torch` instead of `th` throughout the code?

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        val_nid = th.nonzero(g.ndata['val_mask'], as_tuple=True)[0]
+        test_nid = th.nonzero(g.ndata['test_mask'], as_tuple=True)[0]
+        self.train_nid = train_nid
+        self.val_nid = val_nid
+        self.test_nid = test_nid
+
+        # Create sampler
+        sampler = dgl.dataloading.MultiLayerNeighborSampler(
+            [int(fanout) for fanout in args.fan_out.split(',')])
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:104]: See comment below, but I would recommend not accessing `args` directly here, and instead pass everything through `config`.

File: python/ray/util/sgd/torch/examples/gat_dgl.py [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+
+def compute_acc(pred, labels):
+    """
+    Compute the accuracy of prediction given the labels.
+    """
+    return (th.argmax(pred, dim=1) == labels).float().sum() / len(pred)
+
+
+class _NodeCollator(NodeCollator):
```
Comment (amogkam) [inline python/ray/util/sgd/torch/examples/gat_dgl.py:73]: I might be missing something here, but is this necessary? It looks like it's doing the same thing as the superclass

**Review by amogkam — COMMENTED** [review]
Wow this is really cool, thanks for this :)

I left some review comments for better readability and so that it's easier for other people to try out. If you want to add more comments explaining the code, that would help as well.

The last thing is could you make a new directory called `deep_graph` in `torch/examples` and this file to that directory along with a README on how to run this code, what dependencies are required, and what versions these were tested on. You can also include the OMP_NUM_THREADS change you had to make in this README.

After the review comments and the README are addressed, I can merge this in!

File: python/ray/util/sgd/torch/examples/gat_dgl.py (hunk 2 above)
Comment (YeahNew) [reply python/ray/util/sgd/torch/examples/gat_dgl.py:25]: This case is modified from the original gat case of dgl(https://github.com/dmlc/dgl/blob/master/examples/pytorch/ogb/ogbn-products/gat/main.py). The parameter description can be seen by entering the GATConv class, so there is no need to describe it here.

File: python/ray/util/sgd/torch/examples/gat_dgl.py (hunk 9 above)
Comment (YeahNew) [reply python/ray/util/sgd/torch/examples/gat_dgl.py:73]: Thank you for your suggestion, the code has been deleted, and the super class(NodeCollator) is used directly to load data in the custom class(CustomTrainingOperator)

File: python/ray/util/sgd/torch/examples/gat_dgl.py (hunk 7 above)
Comment (YeahNew) [reply python/ray/util/sgd/torch/examples/gat_dgl.py:95]: Ok, I have all modified.

File: python/ray/util/sgd/torch/examples/gat_dgl.py (hunk 6 above)
Comment (YeahNew) [reply python/ray/util/sgd/torch/examples/gat_dgl.py:229]: OK

File: python/ray/util/sgd/torch/examples/gat_dgl.py (hunk 8 above)
Comment (YeahNew) [reply python/ray/util/sgd/torch/examples/gat_dgl.py:104]: OK, All parameters in the custom class have been changed to pass by config.

File: python/ray/util/sgd/torch/examples/gat_dgl.py (hunk 5 above)
Comment (YeahNew) [reply python/ray/util/sgd/torch/examples/gat_dgl.py:268]: OK, I am modifying the code.

**Review by YeahNew — COMMENTED** [review]
> Wow this is really cool, thanks for this :)
> 
> I left some review comments for better readability and so that it's easier for other people to try out. If you want to add more comments explaining the code, that would help as well.
> 
> The last thing is could you make a new directory called `deep_graph` in `torch/examples` and this file to that directory along with a README on how to run this code, what dependencies are required, and what versions these were tested on. You can also include the OMP_NUM_THREADS change you had to make in this README.
> 
> After the review comments and the README are addressed, I can merge this in!

Hi, @amogkam
 Thank you for your valuable comments, I have already modified all of them. Please check if other adjustments are needed.

**Review by amogkam — APPROVED** [review]
